{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d4562d86",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e73ebcff",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "!pip install gen3_validator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b84aa37",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import dictionaryutils as du\n",
    "# dd = du.DataDictionary(root_dir=\"../examples/schema/yaml\")\n",
    "# # getting resolved schema\n",
    "# dd.schema.get(\"lipidomics_file\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4120220d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gen3_validator.resolve_schema import ResolveSchema\n",
    "\n",
    "resolver = ResolveSchema(schema_path=\"../examples/schema/json/acdc_schema.json\")\n",
    "resolver.resolve_schema()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a24bc51",
   "metadata": {},
   "outputs": [],
   "source": [
    "resolver.schema_resolved['demographic.yaml']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05492bd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the node order\n",
    "resolver.node_order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5513b45a",
   "metadata": {},
   "outputs": [],
   "source": [
    "schema = resolver.schema_resolved['proteomics_file.yaml']\n",
    "schema['properties']['data_file_properties.md5sum']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12202d1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# class to extract and store property, data type, and description\n",
    "\n",
    "import logging\n",
    "\n",
    "# Set up logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "schema = resolver.schema_resolved['unaligned_reads_file.yaml']\n",
    "\n",
    "class NodeProps:\n",
    "    def __init__(self, resolved_schema: dict):\n",
    "        self.resolved_schema = resolved_schema\n",
    "    \n",
    "    def get_schema_name(self) -> str:\n",
    "        schema_name = self.resolved_schema['title']\n",
    "        return schema_name\n",
    "    \n",
    "    def get_prop_names(self) -> list:\n",
    "        prop_names = list(self.resolved_schema['properties'].keys())\n",
    "        return prop_names\n",
    "    \n",
    "    def get_data_file_prop_names(self) -> list:\n",
    "        \"\"\"\n",
    "        Returns a list of property names under 'data_file_properties' if it exists,\n",
    "        otherwise returns an empty list.\n",
    "        \"\"\"\n",
    "        prop_names = self.get_prop_names()\n",
    "        if \"data_file_properties\" in prop_names:\n",
    "            return list(self.resolved_schema['properties']['data_file_properties'].keys())\n",
    "        logger.warning(f\"No data_file_properties found in {self.get_schema_name()}\")\n",
    "        return None\n",
    "        \n",
    "\n",
    "    def get_prop_info(self, prop_name: str) -> dict:\n",
    "        prop_names  = self.get_prop_names()\n",
    "        prop_data_file_names = self.get_data_file_prop_names()\n",
    "        prop_info = None\n",
    "\n",
    "        if prop_name in prop_names:\n",
    "            prop_info = self.resolved_schema['properties'][prop_name]\n",
    "        elif prop_data_file_names is not None and prop_name in prop_data_file_names:\n",
    "            prop_info = self.resolved_schema['properties']['data_file_properties'][prop_name]\n",
    "        else:\n",
    "            logger.warning(f\"Property '{prop_name}' not found in {self.get_schema_name()}\")\n",
    "            \n",
    "        return prop_info\n",
    "\n",
    "    def get_data_type(self, prop_name: str) -> str:\n",
    "        prop_info = self.get_prop_info(prop_name)\n",
    "        if prop_info is None:\n",
    "            logger.warning(\n",
    "                f\"Property '{prop_name}' not found in {self.get_schema_name()}, could not pull type\"\n",
    "            )\n",
    "            return None\n",
    "\n",
    "        if \"type\" in prop_info and \"pattern\" in prop_info:\n",
    "            prop_type = f\"string | pattern = {prop_info['pattern']}\"\n",
    "        elif \"type\" in prop_info:\n",
    "            prop_type = prop_info[\"type\"]\n",
    "        elif \"enum\" in prop_info:\n",
    "            prop_type = \"enum\"\n",
    "        else:\n",
    "            logger.warning(\n",
    "                f\"Property '{prop_name}' has no 'type' or 'enum' key. \"\n",
    "                f\"Could be an injected property, usually don't need \"\n",
    "                f\"these in the template | prop_info = {prop_info}\"\n",
    "            )\n",
    "            return None\n",
    "\n",
    "        if not isinstance(prop_type, str):\n",
    "            try:\n",
    "                joined_types = \", \".join(prop_type)\n",
    "                logger.warning(\n",
    "                    f\"Property type '{prop_type}' is not string, converting to string: {joined_types}\"\n",
    "                )\n",
    "                return joined_types\n",
    "            except TypeError:\n",
    "                logger.warning(\n",
    "                    f\"Property type '{prop_type}' is not string and could not be joined.\"\n",
    "                )\n",
    "                return str(prop_type)\n",
    "\n",
    "        return prop_type\n",
    "\n",
    "    def get_description(self, prop_name: str) -> str:\n",
    "        prop_info = self.get_prop_info(prop_name)\n",
    "        if prop_info is None:\n",
    "            logger.warning(f\"\"\"Property '{prop_name}' not found in {self.get_schema_name()}, could not pull description\"\"\"\n",
    "            )\n",
    "            return None\n",
    "        prop_description = None\n",
    "        \n",
    "        if \"description\" in prop_info:\n",
    "            prop_description = prop_info['description']\n",
    "        if \"term\" in prop_info:\n",
    "            prop_description = prop_info['term'][\"description\"]\n",
    "        \n",
    "        if prop_description is None:\n",
    "            logger.warning(f\"\"\"Property '{prop_name}' has no description key. \n",
    "                Could be an injected property, usually don't need these in the\n",
    "                template | prop_info = {prop_info}\"\"\"\n",
    "            )\n",
    "        \n",
    "        return prop_description\n",
    "        \n",
    "# usage\n",
    "nodeprops = NodeProps(schema)\n",
    "# nodeprops.get_description(\"md5sum\")\n",
    "nodeprops.get_data_file_prop_names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d450a0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gen3_metadata_templates.props import NodeProps\n",
    "schema = resolver.schema_resolved['unaligned_reads_file.yaml']\n",
    "nodeprops = NodeProps(schema)\n",
    "nodeprops.get_data_type('md5sum')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55a0d747",
   "metadata": {},
   "outputs": [],
   "source": [
    "nodeprops.get_data_file_prop_names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6073ff88",
   "metadata": {},
   "outputs": [],
   "source": [
    "prop_names = nodeprops.get_prop_names()\n",
    "data_file_prop_names = nodeprops.get_data_file_prop_names()\n",
    "if data_file_prop_names is not None:\n",
    "    prop_names = prop_names.append(data_file_prop_names)\n",
    "prop_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5639d6a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "descriptions = {}\n",
    "for prop_name in prop_names:\n",
    "    descriptions[prop_name] = nodeprops.get_description(prop_name)\n",
    "descriptions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9170f3c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "types = {}\n",
    "for prop_name in prop_names:\n",
    "    types[prop_name] = nodeprops.get_data_type(prop_name)\n",
    "types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7465cf66",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pytest\n",
    "from gen3_validator.resolve_schema import ResolveSchema\n",
    "\n",
    "@pytest.fixture\n",
    "def fixture_res_schema():\n",
    "    resolver = ResolveSchema(schema_path=\"../examples/schema/json/acdc_schema.json\")\n",
    "    resolver.resolve_schema()\n",
    "    return resolver.schema_resolved['unaligned_reads_file.yaml']\n",
    "\n",
    "def test_init_NodeProps(fixture_res_schema):\n",
    "    nodeprops = NodeProps(fixture_res_schema)\n",
    "    assert fixture_res_schema == nodeprops.resolved_schema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab566509",
   "metadata": {},
   "outputs": [],
   "source": [
    "t = \", \".join({\"one\": \"test\", \"two\": \"test\"})\n",
    "t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "41fb60ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:gen3_validator.resolve_schema:Initializing ResolveSchema with schema path: ../examples/schema/json/acdc_schema.json\n",
      "INFO:gen3_validator.resolve_schema:Starting schema resolution process.\n",
      "INFO:gen3_validator.resolve_schema:Reading JSON file from path: ../examples/schema/json/acdc_schema.json\n",
      "INFO:gen3_validator.resolve_schema:Successfully read JSON schema.\n",
      "INFO:gen3_validator.resolve_schema:Retrieving node names from schema.\n",
      "INFO:gen3_validator.resolve_schema:Retrieved 34 nodes from schema.\n",
      "INFO:gen3_validator.resolve_schema:Retrieving all node pairs, excluding specified nodes.\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: demographic.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: demographic.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: project.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: project.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: serum_marker_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: serum_marker_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: alignment_workflow.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: alignment_workflow.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: imaging_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: imaging_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: lipidomics_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: lipidomics_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: metabolomics_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: metabolomics_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: acknowledgement.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: acknowledgement.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: medical_history.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: medical_history.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: blood_pressure_test.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: blood_pressure_test.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: genomics_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: genomics_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: variant_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: variant_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: timepoint.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: timepoint.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: serum_marker_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: serum_marker_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: proteomics_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: proteomics_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: sample.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: sample.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: unaligned_reads_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: unaligned_reads_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: aligned_reads_index_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: aligned_reads_index_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: variant_workflow.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: variant_workflow.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: proteomics_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: proteomics_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: exposure.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: exposure.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: metabolomics_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: metabolomics_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: lipidomics_mapping_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: lipidomics_mapping_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: lipidomics_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: lipidomics_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: aligned_reads_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: aligned_reads_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: lab_result.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: lab_result.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: medication.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: medication.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: publication.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: publication.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: subject.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: subject.yaml\n",
      "INFO:gen3_validator.resolve_schema:Finding upstream and downstream nodes for: core_metadata_collection.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieving links and ID for node: core_metadata_collection.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieved 43 node pairs.\n",
      "INFO:gen3_validator.resolve_schema:Determining node order based on dependencies.\n",
      "INFO:gen3_validator.resolve_schema:Determined node order based on dependencies.\n",
      "INFO:gen3_validator.resolve_schema:Splitting schema into individual node schemas.\n",
      "INFO:gen3_validator.resolve_schema:Split schema into individual node schemas.\n",
      "INFO:gen3_validator.resolve_schema:Retrieving schema for target ID: _definitions.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieved definitions schema.\n",
      "INFO:gen3_validator.resolve_schema:Retrieving schema for target ID: _terms.yaml\n",
      "INFO:gen3_validator.resolve_schema:Retrieved terms schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved references in definitions schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolving all references in schema list.\n",
      "INFO:gen3_validator.resolve_schema:=== Resolving Schema References ===\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved demographic.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved project.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved serum_marker_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved alignment_workflow.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved imaging_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved lipidomics_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved metabolomics_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved acknowledgement.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved medical_history.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved _settings.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved blood_pressure_test.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved genomics_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved variant_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved timepoint.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved program.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved serum_marker_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved proteomics_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved sample.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved unaligned_reads_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved aligned_reads_index_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved variant_workflow.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved proteomics_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved exposure.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved metabolomics_assay.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved lipidomics_mapping_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved lipidomics_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved aligned_reads_file.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved lab_result.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved medication.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved publication.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved subject.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolving references in schema.\n",
      "INFO:gen3_validator.resolve_schema:Resolved core_metadata_collection.yaml\n",
      "INFO:gen3_validator.resolve_schema:Resolved all references in schema list.\n",
      "INFO:gen3_validator.resolve_schema:Converting schema list to JSON format.\n",
      "INFO:gen3_validator.resolve_schema:Converted resolved schema list to JSON format.\n",
      "INFO:gen3_validator.resolve_schema:Obtained schema version: 0.4.7\n",
      "WARNING:gen3_metadata_templates.props:Property type '['string']' is not string, converting to string: string\n",
      "WARNING:gen3_metadata_templates.props:Property 'state' has no 'type' or 'enum' key. Could be an injected property, usually don't need these in the template | prop_info = {'term': {'description': 'The current state of the object.\\n'}, 'default': 'validated', 'downloadable': ['uploaded', 'md5summed', 'validating', 'validated', 'error', 'invalid', 'released'], 'public': ['live'], 'oneOf': [{'enum': ['uploading', 'uploaded', 'md5summing', 'md5summed', 'validating', 'error', 'invalid', 'suppressed', 'redacted', 'live']}, {'enum': ['validated', 'submitted', 'released']}]}\n",
      "WARNING:gen3_metadata_templates.props:Property 'created_datetime' has no 'type' or 'enum' key. Could be an injected property, usually don't need these in the template | prop_info = {'oneOf': [{'type': 'string', 'format': 'date-time'}, {'type': 'null'}], 'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}}\n",
      "WARNING:gen3_metadata_templates.props:Property 'updated_datetime' has no 'type' or 'enum' key. Could be an injected property, usually don't need these in the template | prop_info = {'oneOf': [{'type': 'string', 'format': 'date-time'}, {'type': 'null'}], 'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}}\n",
      "WARNING:gen3_metadata_templates.props:Property 'genomics_assay' has no 'type' or 'enum' key. Could be an injected property, usually don't need these in the template | prop_info = {'anyOf': [{'type': 'array', 'items': {'type': 'object', 'additionalProperties': True, 'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n', 'termDef': {'term': 'Universally Unique Identifier', 'source': 'NCIt', 'cde_id': 'C54100', 'cde_version': None, 'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}}, 'type': 'string', 'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'}, 'submitter_id': {'type': 'string'}}, 'minItems': 1, 'maxItems': 1}}, {'type': 'object', 'additionalProperties': True, 'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n', 'termDef': {'term': 'Universally Unique Identifier', 'source': 'NCIt', 'cde_id': 'C54100', 'cde_version': None, 'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}}, 'type': 'string', 'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'}, 'submitter_id': {'type': 'string'}}}]}\n",
      "WARNING:gen3_metadata_templates.props:Property 'core_metadata_collections' has no 'type' or 'enum' key. Could be an injected property, usually don't need these in the template | prop_info = {'anyOf': [{'type': 'array', 'items': {'type': 'object', 'additionalProperties': True, 'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n', 'termDef': {'term': 'Universally Unique Identifier', 'source': 'NCIt', 'cde_id': 'C54100', 'cde_version': None, 'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}}, 'type': 'string', 'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'}, 'submitter_id': {'type': 'string'}}, 'minItems': 1, 'maxItems': 1}}, {'type': 'object', 'additionalProperties': True, 'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n', 'termDef': {'term': 'Universally Unique Identifier', 'source': 'NCIt', 'cde_id': 'C54100', 'cde_version': None, 'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}}, 'type': 'string', 'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'}, 'submitter_id': {'type': 'string'}}}]}\n",
      "WARNING:gen3_metadata_templates.props:Property 'type' has no description key. Could be an injected property, usually don't need these in the template | prop_info = {'type': 'string'}\n",
      "WARNING:gen3_metadata_templates.props:Property 'genomics_assay' has no description key. Could be an injected property, usually don't need these in the template | prop_info = {'anyOf': [{'type': 'array', 'items': {'type': 'object', 'additionalProperties': True, 'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n', 'termDef': {'term': 'Universally Unique Identifier', 'source': 'NCIt', 'cde_id': 'C54100', 'cde_version': None, 'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}}, 'type': 'string', 'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'}, 'submitter_id': {'type': 'string'}}, 'minItems': 1, 'maxItems': 1}}, {'type': 'object', 'additionalProperties': True, 'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n', 'termDef': {'term': 'Universally Unique Identifier', 'source': 'NCIt', 'cde_id': 'C54100', 'cde_version': None, 'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}}, 'type': 'string', 'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'}, 'submitter_id': {'type': 'string'}}}]}\n",
      "WARNING:gen3_metadata_templates.props:Property 'core_metadata_collections' has no description key. Could be an injected property, usually don't need these in the template | prop_info = {'anyOf': [{'type': 'array', 'items': {'type': 'object', 'additionalProperties': True, 'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n', 'termDef': {'term': 'Universally Unique Identifier', 'source': 'NCIt', 'cde_id': 'C54100', 'cde_version': None, 'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}}, 'type': 'string', 'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'}, 'submitter_id': {'type': 'string'}}, 'minItems': 1, 'maxItems': 1}}, {'type': 'object', 'additionalProperties': True, 'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n', 'termDef': {'term': 'Universally Unique Identifier', 'source': 'NCIt', 'cde_id': 'C54100', 'cde_version': None, 'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}}, 'type': 'string', 'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'}, 'submitter_id': {'type': 'string'}}}]}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['type', 'id', 'submitter_id', 'state', 'project_id', 'created_datetime', 'updated_datetime', 'file_name', 'file_size', 'file_format', 'md5sum', 'object_id', 'file_state', 'error_type', 'ga4gh_drs_uri', 'genomics_assay', 'core_metadata_collections', 'data_category', 'data_format', 'data_type', 'run_id']\n",
      "{'type': 'string', 'id': 'string | pattern = ^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$', 'submitter_id': 'string', 'state': None, 'project_id': 'string', 'created_datetime': None, 'updated_datetime': None, 'file_name': 'string', 'file_size': 'integer', 'file_format': 'string', 'md5sum': 'string | pattern = ^[a-f0-9]{32}$', 'object_id': 'string', 'file_state': 'enum', 'error_type': 'enum', 'ga4gh_drs_uri': 'string', 'genomics_assay': None, 'core_metadata_collections': None, 'data_category': 'enum', 'data_format': 'enum', 'data_type': 'enum', 'run_id': 'string'}\n",
      "{'type': None, 'id': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n', 'submitter_id': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n', 'state': 'The current state of the object.\\n', 'project_id': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n', 'created_datetime': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n', 'updated_datetime': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n', 'file_name': 'The name (or part of a name) of a file (of any type).\\n', 'file_size': 'The size of the data file (object) in bytes.\\n', 'file_format': 'The format of the data file object.\\n', 'md5sum': \"The 128-bit hash value expressed as a 32 digit hexadecimal number used as a file's digital fingerprint.\\n\", 'object_id': 'The GUID of the object in the index service.', 'file_state': 'The current state of the data file object.\\n', 'error_type': 'Type of error for the data file object.\\n', 'ga4gh_drs_uri': 'DRS URI as defined by GA4GH DRS spec for pointers to file objects.\\n', 'genomics_assay': None, 'core_metadata_collections': None, 'data_category': 'Broad categorization of the contents of the data file.', 'data_format': 'Format of the data files.', 'data_type': 'Specific content type of the data file.', 'run_id': 'Sequencing run ID associated with file'}\n"
     ]
    }
   ],
   "source": [
    "from gen3_validator.resolve_schema import ResolveSchema\n",
    "from gen3_metadata_templates.props import NodeProps\n",
    "\n",
    "resolver = ResolveSchema(schema_path=\"../examples/schema/json/acdc_schema.json\")\n",
    "resolver.resolve_schema()\n",
    "\n",
    "node_props = NodeProps(resolver.schema_resolved['unaligned_reads_file.yaml'])\n",
    "\n",
    "\n",
    "node_props.get_schema_name()\n",
    "\n",
    "# return prop names for the schema\n",
    "prop_names = node_props.get_prop_names()\n",
    "print(prop_names)\n",
    "\n",
    "# return the data types for the properties\n",
    "types = {}\n",
    "for prop_name in prop_names:\n",
    "    types[prop_name] = node_props.get_data_type(prop_name)\n",
    "print(types)\n",
    "\n",
    "# return the description for the properties\n",
    "descriptions = {}\n",
    "for prop_name in prop_names:\n",
    "    descriptions[prop_name] = node_props.get_description(prop_name)\n",
    "print(descriptions)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
